XivAlexander/App_Network_SocketHook.cpp

```diff
@@ -222,46 +224,115 @@ class App::Network::SingleConnection::Internals {
 			m_nIoctlTcpInfoFailureCount++;
 			return 0;
-		} else {
+		}
+		else {
 			return std::max(1LL, info.RttUs / 1000LL);
 		}
 	}
 
+	void AddConnectionLatencyItem(int64_t latency) {
+		const int latencyTrackCount = 10;
+
+		ObservedConnectionLatencyList.push_back(latency);
+
+		if (ObservedConnectionLatencyList.size() > latencyTrackCount)
+			ObservedConnectionLatencyList.erase(ObservedConnectionLatencyList.begin());
+	}
+
+	int64_t GetMedianConnectionLatency() const {
+		if (ObservedConnectionLatencyList.empty())
+			return 0;
+
+		size_t count = ObservedConnectionLatencyList.size();
+		std::vector<int64_t> SortedConnectionLatencyList(count);
+		std::partial_sort_copy(ObservedConnectionLatencyList.begin(), ObservedConnectionLatencyList.end(), SortedConnectionLatencyList.begin(), SortedConnectionLatencyList.end());
+
+		if ((count % 2) == 0)
+			// even
+			return (SortedConnectionLatencyList[count / 2] + SortedConnectionLatencyList[count / 2 - 1]) / 2;
+
+		// odd
+		return SortedConnectionLatencyList[count / 2];
+	}
+
+	int64_t GetMeanConnectionLatency() const {
+		if (ObservedServerResponseList.empty())
+			return 0;
+
+		double mean = std::accumulate(ObservedConnectionLatencyList.begin(), ObservedConnectionLatencyList.end(), 0.0) / ObservedConnectionLatencyList.size();
+
+		return std::llround(mean);
+	}
+
+	int64_t GetConnectionLatencyDeviation() const {
+		size_t count = ObservedConnectionLatencyList.size();
+
+		if (count < 2)
+			return 0;
+
+		double mean = std::accumulate(ObservedConnectionLatencyList.begin(), ObservedConnectionLatencyList.end(), 0.0) / count;
+
+		std::vector<double> diff(count);
+		std::transform(ObservedConnectionLatencyList.begin(), ObservedConnectionLatencyList.end(), diff.begin(), [mean](int64_t x) { return double(x) - mean; });
+		double sq_sum = std::inner_product(diff.begin(), diff.end(), diff.begin(), 0.0);
+		double stddev = std::sqrt(sq_sum / count);
+
+		return std::llround(stddev);
+	}
+
 	void AddServerResponseDelayItem(uint64_t delay) {
-		const int latencyTrackCount = ConfigRepository::ConfigRepository().MedianRttCalculationCount;
-		if (latencyTrackCount == 1) {
-			ObservedLatencyList.clear();
-			ObservedLatencyList.push_back(delay);
-			return;
-		}
+		const int latencyTrackCount = 10;
 
-		const auto newPos = ObservedLatencyList.empty() ? 0 : std::upper_bound(ObservedLatencyList.begin(), ObservedLatencyList.end(), delay) - ObservedLatencyList.begin();
-		ObservedLatencyList.insert(ObservedLatencyList.begin() + newPos, delay);
-		if (ObservedLatencyList.size() > latencyTrackCount) {
-			// try to send new item to middle
-			if (newPos < latencyTrackCount / 2)
-				ObservedLatencyList.pop_back();
-			else
-				ObservedLatencyList.erase(ObservedLatencyList.begin());
-		}
+		ObservedServerResponseList.push_back(delay);
+
+		if (ObservedServerResponseList.size() > latencyTrackCount)
+			ObservedServerResponseList.erase(ObservedServerResponseList.begin());
```

Comment:
Since we want to not consider the extreme jitter cases (30, 31, 30, 32, 31, 280, 29, 33, 31, 30), the original code was trying to shift off the extremities at the other end as soon as possible. What would be the reason that it's better to just use all values until a value gets pushed off the list?

Reply:
I actually think server response time tracking is a bit futile shortly after these functions were modified.

But I believe the reason why it was changed was for the sole reason of accuracy relative to time. There was a possibility that repeated high response times would get shifted off and the median would not get updated.

I did use the server response statistic in early development, but it did not seem very reliable.